src/client/Client.cc

```diff
@@ -11029,6 +11029,15 @@ void Client::do_readahead(Fh *f, Inode *in, uint64_t off, uint64_t len)
 
 void Client::C_Read_Async_Finisher::finish(int r)
 {
+  // If inode's file cache cap ref count is more than one then it asserts
+  // ObjectCacher::file_read for async IO returned 0 and file cache cap ref
+  // was incremented. So, it is a must to put it down before readahead
+  // otherwise it will stall the IO call since it will keep on waiting for
+  // the file cache cap to be released.
+  if ((in->cap_refs.at(CEPH_CAP_FILE_CACHE)) > 1) {
```

Comment:
No, this is *absolutely* wrong. In no case can you condition changing a ref count on the current value of the ref count — that just means you're not reference counting correctly elsewhere.
You need to figure out how the other usage is incorrect, and resolve it in a way that is coherent across the entire use system.

Reply:
I do believe the `do_readahead()` code is buggy, because it always assumes that once here it already has gotten the `Frc` caps and will call the `get_cap_ref()` directly instead of `get_caps()`, which will wait for the caps to be available. 

But what if in `Fl` case ? It also could trigger the `do_readahead()` from `_read_async()`. What the kclient will do is that for the readahead it will try to get the needed `Frc` caps just before doing readehead and if fails to get the caps it will just skip the readahead.

Reply:
> But what if in Fl case ? It also could trigger the do_readahead() from _read_async(). What the kclient will do is that for the readahead it will try to get the needed Frc caps just before doing readehead and if fails to get the caps it will just skip the readahead.

We get `Frc` in `_read` so even if it is `Fl` we do have the needed cap ref count while doing readahead 
```
2024-01-25T11:29:26.227+0530 7f5fe588c9c0 10 client.4297 inode before calling _read_async 0x10000000000.head(faked_ino=0 nref=12 ll_ref=1 cap_refs={4=0,1024=1,2048=1,4096=0,8192=1} open={3=1} mode=100666 size=14/4194304 nlink=1 btime=2024-01-25T11:29:26.180449+0530 mtime=2024-01-25T11:29:26.228389+0530 ctime=2024-01-25T11:29:26.228389+0530 change_attr=1 caps=pAsxLsXsxFsxcrwb(0=pAsxLsXsxFsxcrwb) dirty_caps=Fw objectset[0x10000000000 ts 0/0 objects 1 dirty_or_tx 14] parents=0x1.head["test_llreadvllwritevfile475879"] 0x7f5fb80060f0)
```

above is the debug line i had added

Reply:
Please see https://github.com/ceph/ceph/blob/main/src/mds/locks.c#L109-L110, it could just issue `Frl`.

Reply:
https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L10827 here it gets the Frc

Reply:
@dparmar18 As we discussed, when trying to get caps in kclient it will increase all the `got` caps, which is `need | want`, references. 

While for the userspace client it will only increase the `need` caps references. This is why in the `_read_async()` we need to increase the `Fc` caps reference before unlocking the `client_lock` or before doing readahead.

Maybe we can just fix this in `get_caps()` to get the whole references for `need | want` and then it could make the code simpler. 

@gregsfortytwo @vshankar What do you think ?

Reply:
> @dparmar18 As we discussed, when trying to get caps in kclient it will increase all the `got` caps, which is `need | want`, references.
> 
> While for the userspace client it will only increase the `need` caps references. This is why in the `_read_async()` we need to increase the `Fc` caps reference before unlocking the `client_lock` or before doing readahead.
> 
> Maybe we can just fix this in `get_caps()` to get the whole references for `need | want` and then it could make the code simpler.
> 
> @gregsfortytwo @vshankar What do you think ?

If we do this, does this need changes to the way we drop the cap refs? How is it done in the kernel client code?

Reply:
so if the cap ref count of `want` is incremented in `get_caps()`, then the `get_cap_ref()` calls for `want` caps will need to be removed from all the places but this also means that the `put_cap_ref()` calls are still a thing that needs to be managed manually at respective places; operating get and put at different places looks weird and this might be the reason why cap ref count incrementation of `want` caps is handled manually. Need to check how kclient does it

Reply:
I have two solutions in my mind - 
1) As @lxbsz mentioned, enable get_caps() to increment cap ref count of `want` caps too. **OR**
2) Increment cap ref count in _read_async() before calling ObjectCacher::file_read and decrement once we're done reading

Reply:
> @dparmar18 As we discussed, when trying to get caps in kclient it will increase all the `got` caps, which is `need | want`, references.
> 
> While for the userspace client it will only increase the `need` caps references. This is why in the `_read_async()` we need to increase the `Fc` caps reference before unlocking the `client_lock` or before doing readahead.

That is rather an implementation difference IMO: `need` caps _need_ to be issued by the MDS and `want` caps should not be revoked (or under revocation). In the end, the client needs to hold the set of caps that are required for an operation to be performed - so, until the want caps are not released, the functionality is safe, correct? @lxbsz 

> 
> Maybe we can just fix this in `get_caps()` to get the whole references for `need | want` and then it could make the code simpler.

+1 for this.

> 
> @gregsfortytwo @vshankar What do you think ?

(going through further comments on this)

Reply:
> @dparmar18 As we discussed, when trying to get caps in kclient it will increase all the `got` caps, which is `need | want`, references.

When the client does not already have the wanted caps, then only need caps are taken refs on. There is a bit on difference the way ref handling in done b/w the kclient and user-space driver.